prow/tide/tide.go

```diff
@@ -1282,70 +1290,128 @@ func (c *Controller) takeAction(sp subpool, batchPending, successes, pendings, m
 	return Wait, nil, nil
 }
 
-// changedFilesAgent queries and caches the names of files changed by PRs.
-// Cache entries expire if they are not used during a sync loop.
-type changedFilesAgent struct {
-	ghc         githubClient
-	changeCache map[changeCacheKey][]string
-	// nextChangeCache caches file change info that is relevant this sync for use next sync.
-	// This becomes the new changeCache when prune() is called at the end of each sync.
-	nextChangeCache map[changeCacheKey][]string
-	sync.RWMutex
+type gitRepo interface {
+	DiffThreeDot(string, string) ([]string, error)
+	Config(string, string) error
+	Checkout(string) error
+	Merge(string) (bool, error)
+	Clean() error
 }
 
-type changeCacheKey struct {
+type repoCacheKey struct {
 	org, repo string
-	number    int
-	sha       string
 }
 
-// prChanges gets the files changed by the PR, either from the cache or by
-// querying GitHub.
-func (c *changedFilesAgent) prChanges(pr *PullRequest) config.ChangedFilesProvider {
-	return func() ([]string, error) {
-		cacheKey := changeCacheKey{
-			org:    string(pr.Repository.Owner.Login),
-			repo:   string(pr.Repository.Name),
-			number: int(pr.Number),
-			sha:    string(pr.HeadRefOID),
-		}
-
-		c.RLock()
-		changedFiles, ok := c.changeCache[cacheKey]
-		if ok {
-			c.RUnlock()
-			c.Lock()
-			c.nextChangeCache[cacheKey] = changedFiles
-			c.Unlock()
-			return changedFiles, nil
-		}
-		if changedFiles, ok = c.nextChangeCache[cacheKey]; ok {
-			c.RUnlock()
-			return changedFiles, nil
-		}
-		c.RUnlock()
-
-		// We need to query the changes from GitHub.
-		changes, err := c.ghc.GetPullRequestChanges(
-			string(pr.Repository.Owner.Login),
-			string(pr.Repository.Name),
-			int(pr.Number),
+type repoCache interface {
+	getRepo(string, string) (gitRepo, error)
+	// getRepoCopy provides a copy of the repo, which can be used by clients that
+	// may perform destrcutive operations on the repo, e.g. Clean.
+	getRepoCopy(string, string) (gitRepo, error)
+	// prune deletes the repo cache, forcing the fetch of any updates from
+	// remote repositories.
+	prune()
+}
+
+// repoCacher serves a cache of cloned git repos. This is needed to:
+// 1. Avoid calling ClientFor which does a git fetch for updating the repo. With
+//    repoCacher, this update is done only once per sync iteration for each repo.
+// 2. Provide a "throw-away" copy of a cloned repo, which is needed by tide as
+//    it has to delete a repo if it cannot be restored following a failed merge
+//    test.
+type repoCacher struct {
+	cf git.ClientFactory
+
+	// repoCache caches cloned repos during this sync.
+	// This cache is destroyed when prune is called at the end of each sync.
+	repoCache map[repoCacheKey]gitRepo
+	sync.RWMutex
+	copyGitClient git.ClientFactory
+}
+
+func (c *repoCacher) getCopyClient() (git.ClientFactory, error) {
+	if c.copyGitClient == nil {
```

Comment:
I am not sure I understand why this is needed. So we want to call `ClientFor` at most once per repo per sync loop, because it results in actually copying the repo on disk. In order to do that, we cache the `Client` we got via `ClientFor`, that part makes perfect sense. But why is the `copyGitClient` needed?

Reply:
Thank you for the review. This is needed because `Tide` method `pickBatch` destroys the repo:

https://github.com/kubernetes/test-infra/blob/cdd819cd6196ca753210545b24a75663bbfd99e7/prow/tide/tide.go#L965-L969

The `copyGitClient` uses the "main" client factory's cache directory as its source, so it can provide any number of "throw-away" copies of the repo.